queue/QueueConsumer.js

```diff
@@ -24,13 +52,57 @@ module.exports = class QueueConsumer extends BaseQueue {
     if (!this._channel) {
       throw new Error('Channel not initialized')
     }
+
+    const messageContent = message.content
+
+    const maxMessageRetries = this._maxMessageRetries
+    const deadLetterQueueName = this._deadLetterQueueName
+
+    const { headers } = message.properties
+    const deadLetterHeader = headers[DEAD_LETTER_HEADER]
+
+    let rejectedRecord
+    if (deadLetterHeader) {
+      // find 'rejected' record
+      // TODO: handle other reasons in the future
+      // ex. dead letter messages from TTL expires (not implemented yet)
+      for (const record of deadLetterHeader) {
+        if (record.reason === 'rejected') {
+          rejectedRecord = record
+        }
+      }
+    }
+
+    if (rejectedRecord) {
+      // get the number of times this message has been attempted
+      const attemptCount = rejectedRecord.count
+
+      if (attemptCount < maxMessageRetries) {
+        // send to dead letter exchange again
+        return this._channel.nack(message, false, false)
+      }
+
+      // otherwise...
+      // publish to dead letter queue to be handled later
+      await this._channel.sendToQueue(deadLetterQueueName, messageContent)
+
+      // acknowledge message so it doesn't show up on
+      // queue again
+      return this._channel.ack(message)
+    }
+
     // reject message, don't reject all messages,
-    // requeue the message
-    return this._channel.nack(message, false, true)
+    // and don't requeue it in the main queue
+    // (this will move the message over to the dead letter exchange
+    // and will cause the message to be handled by instances of this
+    // consumer again)
+    return this._channel.nack(message, false, false)
```

Comment:
`nack`s now set the `requeue` option to `false`, which will now force messages to be sent to the queue's assigned `dead-letter-exchange` instead of being placed back into it's old position. But, since the exchange is bound to the current queue, the message will still end up being pushed into the consumer's queue. However this time, the built in dead letter extension automatically applies new headers to the message letting us know why the message was placed in the exchange and how many times it has happened.